inc/Search/Indexer.php

```diff
@@ -3,1214 +3,315 @@
 namespace dokuwiki\Search;
 
 use dokuwiki\Extension\Event;
+use dokuwiki\Logger;
+use dokuwiki\Search\Exception\IndexAccessException;
+use dokuwiki\Search\Exception\IndexLockException;
+use dokuwiki\Search\Exception\IndexWriteException;
+
+// Version tag used to force rebuild on upgrade
+const INDEXER_VERSION = 8;
 
 /**
- * Class that encapsulates operations on the indexer database.
+ * Class DokuWiki Indexer
  *
+ * @license    GPL 2 (http://www.gnu.org/licenses/gpl.html)
+ * @author     Andreas Gohr <andi@splitbrain.org>
  * @author Tom N Harris <tnharris@whoopdedo.org>
  */
-class Indexer {
-    /**
-     * @var array $pidCache Cache for getPID()
-     */
-    protected $pidCache = array();
+class Indexer extends AbstractIndex
```

Comment:
To me it makes no sense that this class inherits from AbstractIndex. This muddles the class hierarchy.

* AbstractIndex should define what makes an index (methods for adding, removing, etc)
* FulltextIndex and MetadataIndex should implement specifics inheriting from Abstract Index
* Indexer is not an index in itself. It's a worker that uses the above indexes.